File: src/repositories/mcp-listing-repository.ts
import Database from 'better-sqlite3';
import { McpListing, SearchMcpListingsInput } from '../domain/models.js';

function mapListing(row: Record<string, unknown>): McpListing {
  return {
    id: String(row.id),
    ownerId: String(row.owner_id),
    title: String(row.title),
    summary: String(row.summary),
    version: String(row.version),
    transport: String(row.transport),
    authType: String(row.auth_type),
    sourceType: row.source_type === 'endpoint' ? 'endpoint' : 'package',
    endpointUrl: row.endpoint_url ? String(row.endpoint_url) : null,
    packageName: row.package_name ? String(row.package_name) : null,
    packageRegistry: row.package_registry ? String(row.package_registry) : null,
    sourceUrl: row.source_url ? String(row.source_url) : null,
    homepageUrl: row.homepage_url ? String(row.homepage_url) : null,
    tags: JSON.parse(String(row.tags_json)) as string[],
    taskTypes: JSON.parse(String(row.task_types_json)) as string[],
    isDeleted: Number(row.is_deleted) === 1,
    createdAt: String(row.created_at),
    deletedAt: row.deleted_at ? String(row.deleted_at) : null,
  };
}

export interface CreateListingRecord extends Omit<McpListing, 'tags' | 'taskTypes' | 'isDeleted'> {
  tagsJson: string;
  taskTypesJson: string;
  isDeleted: number;
}

export interface McpListingRepository {
  create(listing: CreateListingRecord): void;
  findById(id: string): McpListing | null;
  softDelete(id: string, deletedAt: string): void;
  searchPublic(input: SearchMcpListingsInput): McpListing[];
  findOwnedByUser(ownerId: string): McpListing[];
}

export function createMcpListingRepository(db: Database.Database): McpListingRepository {
  return {
    create(listing) {
      db.prepare(
        `
          INSERT INTO mcp_listings (
            id, owner_id, title, summary, version, transport, auth_type, source_type,
            endpoint_url, package_name, package_registry, source_url, homepage_url,
            tags_json, task_types_json, is_deleted, created_at, deleted_at
          ) VALUES (
            @id, @ownerId, @title, @summary, @version, @transport, @authType, @sourceType,
            @endpointUrl, @packageName, @packageRegistry, @sourceUrl, @homepageUrl,
            @tagsJson, @taskTypesJson, @isDeleted, @createdAt, @deletedAt
          )
        `,
      ).run(listing);
    },
    findById(id) {
      const row = db.prepare('SELECT * FROM mcp_listings WHERE id = ?').get(id) as Record<string, unknown> | undefined;
      return row ? mapListing(row) : null;
    },
    softDelete(id, deletedAt) {
      db.prepare('UPDATE mcp_listings SET is_deleted = 1, deleted_at = ? WHERE id = ?').run(deletedAt, id);
    },
    searchPublic(input) {
      const conditions = ['is_deleted = 0'];
      const params: unknown[] = [];

      if (input.query) {
        conditions.push('(lower(title) LIKE ? OR lower(summary) LIKE ? OR lower(tags_json) LIKE ? OR lower(task_types_json) LIKE ?)');
        const likeValue = `%${input.query}%`;
        params.push(likeValue, likeValue, likeValue, likeValue);
      }

      if (input.taskType) {
        conditions.push('lower(task_types_json) LIKE ?');
        params.push(`%${input.taskType}%`);
      }

      if (input.transport) {
        conditions.push('transport = ?');
        params.push(input.transport);
      }

      const rows = db
        .prepare(`SELECT * FROM mcp_listings WHERE ${conditions.join(' AND ')} ORDER BY created_at DESC`)
        .all(...params) as Record<string, unknown>[];

      return rows.map(mapListing);
    },
    findOwnedByUser(ownerId) {
      const rows = db
        .prepare('SELECT * FROM mcp_listings WHERE owner_id = ? AND is_deleted = 0 ORDER BY created_at DESC')
        .all(ownerId) as Record<string, unknown>[];

      return rows.map(mapListing);
    },
  };
}

File: src/repositories/mcp_listing_repository.py
import json
import sqlite3

from src.domain.models import McpListing, SearchMcpListingsInput


def _map_listing(row: sqlite3.Row) -> McpListing:
    return McpListing(
        id=str(row["id"]),
        owner_id=str(row["owner_id"]),
        title=str(row["title"]),
        summary=str(row["summary"]),
        version=str(row["version"]),
        transport=str(row["transport"]),
        auth_type=str(row["auth_type"]),
        source_type=str(row["source_type"]),
        endpoint_url=str(row["endpoint_url"]) if row["endpoint_url"] else None,
        package_name=str(row["package_name"]) if row["package_name"] else None,
        package_registry=str(row["package_registry"]) if row["package_registry"] else None,
        source_url=str(row["source_url"]) if row["source_url"] else None,
        homepage_url=str(row["homepage_url"]) if row["homepage_url"] else None,
        tags=json.loads(str(row["tags_json"])),
        task_types=json.loads(str(row["task_types_json"])),
        is_deleted=int(row["is_deleted"]) == 1,
        created_at=str(row["created_at"]),
        deleted_at=str(row["deleted_at"]) if row["deleted_at"] else None,
    )


class McpListingRepository:
    def __init__(self, connection: sqlite3.Connection) -> None:
        self._connection = connection

    def create(self, listing: McpListing) -> None:
        self._connection.execute(
            """
            INSERT INTO mcp_listings (
              id, owner_id, title, summary, version, transport, auth_type, source_type,
              endpoint_url, package_name, package_registry, source_url, homepage_url,
              tags_json, task_types_json, is_deleted, created_at, deleted_at
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                listing.id,
                listing.owner_id,
                listing.title,
                listing.summary,
                listing.version,
                listing.transport,
                listing.auth_type,
                listing.source_type,
                listing.endpoint_url,
                listing.package_name,
                listing.package_registry,
                listing.source_url,
                listing.homepage_url,
                json.dumps(listing.tags, ensure_ascii=False),
                json.dumps(listing.task_types, ensure_ascii=False),
                1 if listing.is_deleted else 0,
                listing.created_at,
                listing.deleted_at,
            ),
        )
        self._connection.commit()

    def find_by_id(self, listing_id: str) -> McpListing | None:
        row = self._connection.execute("SELECT * FROM mcp_listings WHERE id = ?", (listing_id,)).fetchone()
        return _map_listing(row) if row else None

    def soft_delete(self, listing_id: str, deleted_at: str) -> None:
        self._connection.execute(
            "UPDATE mcp_listings SET is_deleted = 1, deleted_at = ? WHERE id = ?",
            (deleted_at, listing_id),
        )
        self._connection.commit()

    def search_public(self, search_input: SearchMcpListingsInput) -> list[McpListing]:
        conditions = ["is_deleted = 0"]
        params: list[str] = []

        if search_input.query:
            conditions.append(
                "(lower(title) LIKE ? OR lower(summary) LIKE ? OR lower(tags_json) LIKE ? OR lower(task_types_json) LIKE ?)"
            )
            like_value = f"%{search_input.query}%"
            params.extend([like_value, like_value, like_value, like_value])

        if search_input.task_type:
            conditions.append("lower(task_types_json) LIKE ?")
            params.append(f"%{search_input.task_type}%")

        if search_input.transport:
            conditions.append("transport = ?")
            params.append(search_input.transport)

        rows = self._connection.execute(
            f"SELECT * FROM mcp_listings WHERE {' AND '.join(conditions)} ORDER BY created_at DESC",
            tuple(params),
        ).fetchall()
        return [_map_listing(row) for row in rows]

    def find_owned_by_user(self, owner_id: str) -> list[McpListing]:
        rows = self._connection.execute(
            "SELECT * FROM mcp_listings WHERE owner_id = ? AND is_deleted = 0 ORDER BY created_at DESC",
            (owner_id,),
        ).fetchall()
        return [_map_listing(row) for row in rows]
